torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -120,33 +124,205 @@ def can_auto_functionalize(op: OperatorBase) -> bool:
     _mutable_op: OpOverload,
     _only_clone_these_tensors: Optional[Tuple[str, ...]] = None,
     **kwargs: Any,
 ) -> Tuple[Any, Tuple[Tensor, ...]]:
+    all_bases: List[Tensor] = kwargs.pop("_all_bases", [])
+    mutable_args_names, mutable_args_types = get_mutable_args(_mutable_op)
+    args_view_info = deserialize_views_meta(
+        mutable_args_names, mutable_args_types, kwargs, all_bases, pop_args=True
+    )
+
+    def regenerate_view(ViewInfo):
+        if (
+            ViewInfo.base.size() == ViewInfo.size
+            and ViewInfo.base.stride() == ViewInfo.stride
+            and ViewInfo.base.storage_offset() == ViewInfo.storage_offset
+        ):
+            return ViewInfo.base
+        return torch.as_strided(
+            ViewInfo.base, ViewInfo.size, ViewInfo.stride, ViewInfo.storage_offset
+        )
+
+    # Re-generate all inputs from _all_bases using args_view_info and add them to kwargs.
+    for arg_name in mutable_args_names:
+        if args_view_info[arg_name] is None:
+            kwargs[arg_name] = None
+        elif isinstance(args_view_info[arg_name], list):
+            kwargs[arg_name] = []
+            for i, elem in enumerate(args_view_info[arg_name]):
+                if elem is None:
+                    kwargs[arg_name].append(None)
+                else:
+                    view_info = args_view_info[arg_name][i]
+                    kwargs[arg_name].append(regenerate_view(view_info))
+        else:
+            kwargs[arg_name] = regenerate_view(args_view_info[arg_name])
+
+    # create new args
     new_kwargs = dict(**kwargs)
     result = []
 
-    _mutable_args_names = get_mutable_arg_names(_mutable_op)
-    for name in _mutable_args_names:
+    for name in mutable_args_names:
         if (
             _only_clone_these_tensors is not None
             and name not in _only_clone_these_tensors
         ):
             new_kwargs[name] = kwargs[name]
         else:
-            new_kwargs[name] = (
-                [clone_preserve_strides(x) for x in kwargs[name]]
-                if kwargs[name] is not None and isinstance(kwargs[name], list)
-                else clone_preserve_strides(kwargs[name])
-                if kwargs[name] is not None
-                else None
-            )
-        result.append(new_kwargs[name])
+            if kwargs[name] is not None and isinstance(kwargs[name], list):
+                new_kwargs[name] = [clone_preserve_strides(x) for x in kwargs[name]]
+
+            elif kwargs[name] is not None:
+                new_kwargs[name] = clone_preserve_strides(kwargs[name])
+
     out = _mutable_op(**new_kwargs)
 
+    def observe_mutation(base, mutation_source):
+        if (
+            mutation_source.size() == base.size()
+            and mutation_source.stride() == base.stride()
+            and mutation_source.storage_offset() == base.storage_offset()
+        ):
+            return mutation_source
+
+        return base.as_strided_scatter(
```

Comment:
Do we need as_strided_scatter? Alternative, auto_functionalized can:
1) clone the bases preserving strides (if necessary)
2) call as_strided on them to reconstruct the views
3) run the mutable op.
3) return the (possibly cloned) bases preserving strides.

I think this would also greatly simplify the logic in this function

Reply:
addressed this simpler